File: cpp/fivecardstud.cpp
#include <sstream>
#include <iostream>
#include <fstream>
#include <vector>
#include <set>
#include <map>
#include <algorithm>
#include "card.h"
#include "deck.h"

using namespace std;

// enumeration for suits
enum Suit {
	Diamonds, Clubs, Hearts, Spades
};

// enumeration for hand ranks 
enum HandRank {
	HIGH_CARD,
	PAIR,
	TWO_PAIR,
	THREE_OF_A_KIND,
	STRAIGHT,
	FLUSH,
	FULL_HOUSE,
	FOUR_OF_A_KIND,
	STRAIGHT_FLUSH,
	ROYAL_FLUSH
};

// method headers 
HandRank classifyHand(vector<Card> hand);
bool isStraight(const vector<Card>& hand);
bool isAceLowStraight(const vector<Card>& hand);
Card getHighestCard(const vector<Card>& hand, bool isAceStraight);
int countOccurrences(const vector<Card>& hand, int rank);
Card getHighestNonPairCard(const vector<Card>& hand);
int getKicker(const vector<Card>& hand);
int getPairRank(const vector<Card>& hand);
int getHighestPairRank(const vector<Card>& hand);
int getLowestPairRank(const vector<Card>& hand);
int getTripletRank(const vector<Card>& hand);
int getQuadRank(const vector<Card>& hand);
bool compareHands(const vector<Card>& hand1, const vector<Card>& hand2);
void sortHands(vector<vector<Card>>& hands, vector<HandRank>& handRanks);
void printSixHands(const vector<vector<Card>>& hands);
void printWinningHand(const vector<vector<Card>>& hands, const vector<HandRank>& handRanks);
void readDeckFromFile(const string& filename, vector<vector<Card>>& hands);


// main method
int main(int argc, char* argv[]) {
	Deck deck; // create ordered deck of cards 
	vector<vector<Card>> hands(6); // to store the 6 hands
	vector<HandRank> handRanks; // to store ranks of each hand 
	
	cout << "*** P O K E R    H A N D    A N A L Y Z E R ***" << endl;
	cout << endl;
	cout << endl;

	if(argc > 1) { // part 2 w/ command line args
		cout << "*** USING TEST DECK ***" << endl;
		cout << endl;
		cout << "*** File: " << argv[1] << endl;
		readDeckFromFile(argv[1], hands);
	} else { // part 1 w/o command line args 
		cout << "*** USING RANDOMIZED DECK OF CARDS ***" << endl;
		cout << endl;
		cout << "*** Shuffled 52 card deck:" << endl;
                deck.shuffle();
                deck.printDeck(false);
                deck.dealHands(hands);
                printSixHands(hands);
                cout << "*** Here is what remains in the deck..." << endl;
                deck.printDeck(true);
                cout << endl;
	}
	sortHands(hands, handRanks);
        printWinningHand(hands, handRanks);
        return 0;
}

// map hand rank enum values to strings 
map<int, string> handRankToString = {
	{HIGH_CARD, "High Card"},
	{PAIR, "Pair"},
	{TWO_PAIR, "Two Pair"},
	{THREE_OF_A_KIND, "Three Of A Kind"},
	{STRAIGHT, "Straight"},
	{FLUSH, "Flush"},
	{FULL_HOUSE, "Full House"},
	{FOUR_OF_A_KIND, "Four Of A Kind"},
	{STRAIGHT_FLUSH, "Straight Flush"},
	{ROYAL_FLUSH, "Royal Flush"},
};

// checks if the hand is a straight
bool isStraight(const vector<Card>& hand) {  
	vector<int> ranks;
	for(int i = 0; i < hand.size(); i++) {
		ranks.push_back(hand[i].rank);
	}

	sort(ranks.begin(), ranks.end()); // ascending order 

	// special check for Ace-low straight (A, 2, 3, 4, 5)
	if(ranks == vector<int>{2,3,4,5,14}) {
		return true;
	}

	// checks if ranks are consecutive
	for(int i = 0; i < ranks.size() - 1; i++) { 
		if(ranks[i + 1] != ranks[i] + 1) {
			return false;
		}
	}
	return true;
}

// compare cards by rank in descending order 
bool compareCardsByRank(const Card& a, const Card& b) {
	return a.rank > b.rank;	
}

// printing method to display 6 hands 
void printSixHands(const vector<vector<Card>>& hands) {
        cout << "*** Here are the six hands..." << endl;
        for (int i = 0; i < 6; i++) {
                for (int j = 0; j < hands[i].size(); j++) {
                        cout << hands[i][j].toString() << " ";
                }
                cout << endl;
        }
	cout << endl;
}

// printing method to display winning hand order 
void printWinningHand(const vector<vector<Card>>& hands, const vector<HandRank>& handRanks) {
	cout << "--- WINNING HAND ORDER ---" << endl;
        for(int i = 0; i < hands.size(); i++) {
                for(int j = 0; j < hands[i].size(); j++) {
                        cout << hands[i][j].toString() << " ";
                }
                cout << "- " << handRankToString[handRanks[i]] << endl;
        }
}

// determines rank if hand based on card combinations 
HandRank classifyHand(vector<Card> hand) { 
	sort(hand.begin(), hand.end(), compareCardsByRank); // descending order e.g 14, 13, 12, 11, 10 

	// to count occurences of each value 
	int occurences[15] = {0}; 
	int suitCount[4] = {0};;
	const char suits[] = {'D', 'C', 'H', 'S'};

	for(int i = 0; i < hand.size(); i++) {
		occurences[hand[i].rank]++; // count rank occurences 

		for(int j = 0; j < 4; j++) {
			if(hand[i].suit == suits[j]) {
				suitCount[j]++; // count suit occurences 
				break;
			}
		}
	}

	bool flush = false;
	for(int i = 0; i < 4; i++) { // checks if hand is a flush (same suit)
		if(suitCount[i] == 5) {
			flush = true;
			break;
		}
	}
	bool straight = isStraight(hand);

	if(flush && straight) {
		set<int> royalRanks = {14, 13, 12, 11, 10};
		set<int> uniqueRanks;
		for (int i = 0; i < hand.size(); i++) {
			uniqueRanks.insert(hand[i].rank);
		}
		
		if(uniqueRanks == royalRanks) {
			return ROYAL_FLUSH;
		} else {
			return STRAIGHT_FLUSH;
		}

	} 

	int threeCount = 0;
	int pairCount = 0;
	for(int i = 2; i <= 14; i++) {
		if(occurences[i] == 4) return FOUR_OF_A_KIND;
		if(occurences[i] == 3) threeCount++;
		if(occurences[i] == 2) pairCount++;
	}

	if(threeCount == 1 && pairCount == 1) return FULL_HOUSE; // working 
	if(flush) return FLUSH; // working 
	if(straight) return STRAIGHT; // working 
	if(threeCount == 1) return THREE_OF_A_KIND; // working 
	if(pairCount == 2) return TWO_PAIR; // working 
	if(pairCount == 1) return PAIR; // working 

	return HIGH_CARD;

}

// check if a hand is an Ace-low straight
bool isAceLowStraight(const vector<Card>& hand) {
	for(int i = 0; i < hand.size(); i++) {
		if(hand[i].rank == 14) {
			return true; 
		}
	}
	return false;
}

// get highest number card, includes case of Ace-low straight 
Card getHighestCard(const vector<Card>& hand, bool isAceStraight) {
	vector<Card> tempHand = hand;
	
	if (isAceStraight) {
		for (int i = 0; i < tempHand.size(); ++i) {
			if (tempHand[i].rank == 14) {
				tempHand[i].rank = 1; // treat Ace as a low 
			}
		}
	}

	Card highestCard = tempHand[0];

	for (int i = 1; i < tempHand.size(); i++) {
		if (tempHand[i].rank > highestCard.rank) { 
			highestCard = tempHand[i];
		} 
	}
	return highestCard;

}

// count occurrences of a specific rank in a hand
int countOccurrences(const vector<Card>& hand, int rank) {
	int count = 0;
	for(int i = 0; i < hand.size(); i++) {
		if(hand[i].rank == rank) {
			count++;
		}
	}
	return count;
}

// get the highest card not part of a pair
Card getHighestNonPairCard(const vector<Card>& hand) {
	vector<Card> tempHand;
	
	// add cards that are not part of pair to temphand 
	for(int i = 0; i < hand.size(); i++) {
		if((countOccurrences(hand, hand[i].rank) != 2)) {
			tempHand.push_back(hand[i]);
		}
	}

	Card highestCard = tempHand[0];
	for (int i = 1; i < tempHand.size(); i++) {
		if (tempHand[i].rank > highestCard.rank) {
			highestCard = tempHand[i];
		}
	}
	return highestCard;
}

// identify kicker card in a hand (not part of pair)
int getKicker(const vector<Card>& hand) {
	int maxVal = 0;
	for(int i = 0; i < hand.size(); i++) {
		if((countOccurrences(hand, hand[i].rank) == 1) && (hand[i].rank > maxVal)) {
			maxVal = hand[i].rank;
		}
	}
	return maxVal;
}

// get the rank (value) of the pair in a hand
int getPairRank(const vector<Card>& hand) {
	for(int i = 0; i < hand.size(); i++) {
		if(countOccurrences(hand, hand[i].rank) == 2) {
			return hand[i].rank;
		}
	}
	return -1;
}

// get the highest rank (value) of pairs in a hand
int getHighestPairRank(const vector<Card>& hand) {
	int highestPair = -1;
	for(int i = 0; i < hand.size(); i++) {
		if(countOccurrences(hand, hand[i].rank) == 2) {
			if(hand[i].rank > highestPair) {
				highestPair = hand[i].rank;
			}
		}
	}
	return highestPair;
}

// get the lowest rank (value) of pairs in a hand
int getLowestPairRank(const vector<Card>& hand) {
	int lowestPair = 15;
	for(int i = 0; i < hand.size(); i++) {
		if(countOccurrences(hand, hand[i].rank) == 2) {
			if(hand[i].rank < lowestPair) {
				lowestPair = hand[i].rank;
			}	
		}
	}
	return lowestPair;
}

// get the rank of the three-of-a-kind in a hand
int getTripletRank(const vector<Card>& hand) {
	for(int i = 0; i < hand.size(); i++) {
		if(countOccurrences(hand, hand[i].rank) == 3) {
			return hand[i].rank;
		}
	}
	return -1;
}

// get the rank of the four-of-a-kind in a hand
int getQuadRank(const vector<Card>& hand) {
	for(int i = 0; i < hand.size(); i++) {
		if(countOccurrences(hand, hand[i].rank) == 4) {
			return hand[i].rank;
		}
	}
	return -1;
}

// compare two hands to determine the winner
bool compareHands(const vector<Card>& hand1, const vector<Card>& hand2) {
	HandRank rank1 = classifyHand(hand1);
	HandRank rank2 = classifyHand(hand2);
	
	// first, compare overall hand ranks 
	if(rank1 != rank2) {
		return rank1 > rank2;
	}

	// tie-breaking logic for equal strength hands  
	
	// case 1 flush categories - use suit rank to break ties (lowest to highest: D C H S) 
	if(rank1 == FLUSH || rank1 == STRAIGHT_FLUSH || rank1 == ROYAL_FLUSH) {
		Card highestCard1 = getHighestCard(hand1, false);
		Card highestCard2 = getHighestCard(hand2, false);

		if(highestCard1.rank != highestCard2.rank) {
			return highestCard1.rank > highestCard2.rank;
		}	
		
		return highestCard1.suit > highestCard2.suit; // if identical 
			
	}	

	// case 2 straights but not flushes - use suit rank to break ties
	if (rank1 == STRAIGHT) {
		bool checkAce1 = isAceLowStraight(hand1);
		bool checkAce2 = isAceLowStraight(hand2);

		Card highestCard1 = getHighestCard(hand1, checkAce1);
		Card highestCard2 = getHighestCard(hand2, checkAce2);

		if(highestCard1.rank != highestCard2.rank) {
			return highestCard1.rank > highestCard2.rank;
		}

		return highestCard1.suit > highestCard2.suit; // if identical
	}	

	// case 3 - two pair - suit of kicker card to break ties based on (1) 
	if(rank1 == TWO_PAIR) {
		int highPair1 = getHighestPairRank(hand1);
		int highPair2 = getHighestPairRank(hand2);
		
		if(highPair1 != highPair2) {
			return highPair1 > highPair2;
		}
	
		int lowPair1 = getLowestPairRank(hand1);
		int lowPair2 = getLowestPairRank(hand2);
		
		if(lowPair1 != lowPair2) {
			return lowPair1 > lowPair2;
		}

		// if identical, compare kicker suit 
		int kicker1 = getKicker(hand1);
		int kicker2 = getKicker(hand2);
		return hand1[kicker1].suit > hand2[kicker2].suit;	
	}

	// case 4 pairs - use highest non-pair card to break tie 
	if (rank1 == PAIR) {
		int pairRank1 = getPairRank(hand1);
		int pairRank2 = getPairRank(hand2);

		if(pairRank1 != pairRank2) {
			return pairRank1 > pairRank2;
		}

		Card highVal1 = getHighestNonPairCard(hand1);
		Card highVal2 = getHighestNonPairCard(hand2);

		if(highVal1.rank != highVal2.rank) {
			return highVal1.rank > highVal2.rank;
		}
		return highVal1.suit > highVal2.suit; // if identical
	}	

	// case 5 high card - suit of highest value hard to break ties 
	if(rank1 == HIGH_CARD) {
		Card highestCard1 = getHighestCard(hand1, false);
		Card highestCard2 = getHighestCard(hand2, false);

		if(highestCard1.rank != highestCard2.rank) {
			return highestCard1.rank > highestCard2.rank;
		}
		return highestCard1.suit > highestCard2.suit; // if identical
	}

	if(rank1 == FOUR_OF_A_KIND) { // directly compare quadranks 
		int quadRank1 = getQuadRank(hand1);
		int quadRank2 = getQuadRank(hand2);

		return quadRank1 > quadRank2;
	}

	if(rank1 == THREE_OF_A_KIND) { // directly compare triplet ranks
		int tripletRank1 = getTripletRank(hand1);
		int tripletRank2 = getTripletRank(hand2);
		return tripletRank1 > tripletRank2;
	}	

	if(rank1 == FULL_HOUSE) { // compare triplet ranks first, then by pair if equal
		int tripletRank1 = getTripletRank(hand1);
		int tripletRank2 = getTripletRank(hand2);

		if(tripletRank1 != tripletRank2) {
			return tripletRank1 > tripletRank2;
		}

		int pairRank1 = getPairRank(hand1);
		int pairRank2 = getPairRank(hand2);

		return pairRank1 > pairRank2;
	}

	return false;
}

// sort hands based on rank 
void sortHands(vector<vector<Card>>& hands, vector<HandRank>& handRanks) {
	sort(hands.begin(), hands.end(), compareHands);
	for (int i = 0; i < hands.size(); ++i) {
		HandRank rank = classifyHand(hands[i]);
		handRanks.push_back(rank);
	}
}

// read deck from file and populate hands
void readDeckFromFile(const string& filename, vector<vector<Card>>& hands) {
	hands.resize(6);
	ifstream inputFile(filename); // open file 
	string line;
	set<string> seenCards;

	vector<string> lines;
	while (getline(inputFile, line)) {
		lines.push_back(line);
		cout << line << endl; // display file line to user 
	}	

	for(int i = 0; i < 6; i++) { 
		stringstream ss(lines[i]);
		string cardStr;

		for(int j = 0; j < 5; j++) {
			if(j < 4) { // for the first four cards, read until comma 
				if (!getline(ss, cardStr, ',')) {
					cerr << "Error: Invalid card format in file." << endl;
					exit(1);
				}
			} else {
				if (!getline(ss, cardStr)) {
					cerr << "Error: Invalid card format in file." << endl;
					exit(1);
				}
			}
			
			cardStr.erase(remove(cardStr.begin(), cardStr.end(), ' '), cardStr.end());
			
			
			// duplicate card check 
			if(seenCards.find(cardStr) != seenCards.end()) {
				cout << endl;
				cout << "*** ERROR - DUPLICATED CARD FOUND IN DECK ***" << endl;
				cout << endl;
				cout << endl;
				cout << "*** DUPLICATE: " << cardStr << " ***" << endl;
				exit(1);
			}
			seenCards.insert(cardStr);

			char suit = cardStr.back();
			int rank = 0;

			// card conversion 
			if (cardStr[0] >= '2' && cardStr[0] <= '9') {
				rank = cardStr[0] - '0';
			} else if (cardStr[0] == '1' && cardStr[1] == '0') {
				rank = 10;
			} else {
				switch (cardStr[0]) {
					case 'J': rank = 11; break; 
					case 'Q': rank = 12; break;
					case 'K': rank = 13; break;
					case 'A': rank = 14; break;
				
				}
			}
	
			hands[i].push_back(Card(rank, suit));
		}
	}
	cout << endl;
	printSixHands(hands);
}

File: julia/FiveCardStud.jl
# FiveCardStud.jl

include("Card.jl")
include("Deck.jl")

using .DeckModule
using .CardModule

# Define HandRank as an Enum
@enum HandRank HIGH_CARD PAIR TWO_PAIR THREE_OF_A_KIND STRAIGHT FLUSH FULL_HOUSE FOUR_OF_A_KIND STRAIGHT_FLUSH ROYAL_FLUSH

# Function to map hand ranks to strings
handRankToString = Dict(
    	HIGH_CARD => "High Card",
    	PAIR => "Pair",
    	TWO_PAIR => "Two Pair",
    	THREE_OF_A_KIND => "Three Of A Kind",
    	STRAIGHT => "Straight",
    	FLUSH => "Flush",
    	FULL_HOUSE => "Full House",
    	FOUR_OF_A_KIND => "Four Of A Kind",
    	STRAIGHT_FLUSH => "Straight Flush",
    	ROYAL_FLUSH => "Royal Straight Flush"
)

# create Dict for suit ranking lowest to highest: Diamonds, Clubs, Hearts, Spades.
function suit_rank(suit::Char)::Int
    	return Dict('D' => 1, 'C' => 2, 'H' => 3, 'S' => 4)[suit]
end

# Function to print the six hands of cards
function print_six_hands(hands::Vector{Vector{Card}})
	println("*** Here are the six hands...")
	for (i, hand) in enumerate(hands)
		println(join([to_string(card) for card in hand], " "))
	end
	println()
end

# Function to print winning hand order with hand ranks
function print_winning_hand(hands::Vector{Vector{Card}}, handRanks::Vector{HandRank})
    	println("--- WINNING HAND ORDER ---")
    	for i in 1:length(hands)
        	println(join([to_string(card) for card in hands[i]], " "), " - ", handRankToString[handRanks[i]])
    	end
end

# Sort hands based on rank
function sort_hands(hands::Vector{Vector{Card}}, handRanks::Vector{HandRank})
	sorted_hands = sort(hands, lt = (hand1, hand2) -> compare_hands(hand1, hand2))
	empty!(handRanks)
	
	for hand in sorted_hands
        	rank = classify_hand(hand)
        	push!(handRanks, rank)
    	end

    	return sorted_hands
end

# Function to check if a hand is a straight
function is_straight(hand::Vector{Card})
	#temp_hand = sort([card for card in hand], by = x -> x.rank)
	ranks = [card.rank for card in hand]
	sorted_ranks = sort(ranks)

	# Special check for Ace-low straight (A, 2, 3, 4, 5)
	if sorted_ranks == [2, 3, 4, 5, 14]
		return true
	end
	
	for i in 1:(length(sorted_ranks) - 1)
        	if sorted_ranks[i + 1] != sorted_ranks[i] + 1
            		return false
        	end
    	end

    	return true
end

# Check if a hand is an Ace-low straight
function is_ace_low_straight(hand::Vector{Card})::Bool
	for card in hand
        	if card.rank == 14
            		return true  # Ace is present
        	end
    	end
    	return false
end

# Get the highest card, includes case of Ace-low straight
function get_highest_card(hand::Vector{Card}, isAceStraight::Bool)::Card
	highest_card = hand[1]
    	
	# Treat Ace as 1 if isAceStraight is true

	if isAceStraight && highest_card.rank == 14
        	highest_rank = 1
    	else
        	highest_rank = highest_card.rank
    	end

    	for i in 2:length(hand)
        	current_card = hand[i]
        	if isAceStraight && current_card.rank == 14
            		current_rank = 1
        	else
            		current_rank = current_card.rank
        	end

        	if current_rank > highest_rank
            		highest_card = current_card
            		highest_rank = current_rank
        	end
    	end
    	return highest_card
end

# Count occurrences of a specific rank in a hand
function count_occurrences(hand::Vector{Card}, rank::Int)::Int
	count = 0
    	for card in hand
        	if card.rank == rank
            		count += 1
        	end
    	end
    	return count
end

# Get the highest card not part of a pair
function get_highest_non_pair_card(hand::Vector{Card})::Card
	temp_hand = Card[]
    
    	for card in hand
        	if count_occurrences(hand, card.rank) != 2
            		push!(temp_hand, card)
        	end
    	end

    	highest_card = temp_hand[1]
    	for i in 2:length(temp_hand)
        	if temp_hand[i].rank > highest_card.rank
            		highest_card = temp_hand[i]
        	end
    	end
    
	return highest_card
end

# Identify index of kicker card in a hand (not part of pair)
function get_kicker(hand::Vector{Card})::Int
	max_val = 0
	kicker_index = -1

	for i in 1:length(hand)
		card = hand[i]
		if count_occurrences(hand, card.rank) == 1 && card.rank > max_val
			max_val = card.rank
			kicker_index = i
		end
	end
	
	return kicker_index
end

# Get the rank of the pair in a hand
function get_pair_rank(hand::Vector{Card})::Int
	for card in hand
        	if count_occurrences(hand, card.rank) == 2
            		return card.rank
        	end
    	end
    	return -1
end

# Get the highest rank of pairs in a hand
function get_highest_pair_rank(hand::Vector{Card})::Int
	highest_pair = -1
    	for card in hand
        	if count_occurrences(hand, card.rank) == 2 && card.rank > highest_pair
            		highest_pair = card.rank
        	end
    	end
    	return highest_pair
end

# Get the lowest rank of pairs in a hand
function get_lowest_pair_rank(hand::Vector{Card})::Int
	lowest_pair = 15
    	for card in hand
        	if count_occurrences(hand, card.rank) == 2 && card.rank < lowest_pair
            		lowest_pair = card.rank
        	end
    	end
    	return lowest_pair
end

# Get the rank of the three-of-a-kind in a hand
function get_triplet_rank(hand::Vector{Card})::Int
	for card in hand
        	if count_occurrences(hand, card.rank) == 3
            		return card.rank
        	end
    	end
    	return -1
end

# Get the rank of the four-of-a-kind in a hand
function get_quad_rank(hand::Vector{Card})::Int
	for card in hand
        	if count_occurrences(hand, card.rank) == 4
            		return card.rank
        	end
    	end
    	return -1
end

# to classify a hand with its HandRank
function classify_hand(hand::Vector{Card})::HandRank
	sort(hand, by=x -> x.rank, rev=true)  # Sort in descending order

	occurrences = fill(0, 15)
	suit_count = Dict('D' => 0, 'C' => 0, 'H' => 0, 'S' => 0)

	for card in hand
		occurrences[card.rank] += 1
		suit_count[card.suit] += 1
	end

	flush = any(count == 5 for count in values(suit_count))
	straight = is_straight(hand)

	if flush && straight
		royal_ranks = Set([14, 13, 12, 11, 10])
		hand_ranks = Set([card.rank for card in hand])

		if hand_ranks == royal_ranks
			return ROYAL_FLUSH
		else
			return STRAIGHT_FLUSH
		end
	end

	three_count = 0
    	pair_count = 0
    	for rank in 2:14
        	if occurrences[rank] == 4
            		return FOUR_OF_A_KIND
        	elseif occurrences[rank] == 3
            		three_count += 1
        	elseif occurrences[rank] == 2
            		pair_count += 1
        	end
    	end

    	if three_count == 1 && pair_count == 1
        	return FULL_HOUSE
    	elseif flush
        	return FLUSH
    	elseif straight
        	return STRAIGHT
    	elseif three_count == 1
        	return THREE_OF_A_KIND
    	elseif pair_count == 2
        	return TWO_PAIR
    	elseif pair_count == 1
        	return PAIR
    	else
        	return HIGH_CARD
    	end
end

# compare two hands to determine the winner
function compare_hands(hand1::Vector{Card}, hand2::Vector{Card})::Bool
	rank1 = classify_hand(hand1)
	rank2 = classify_hand(hand2)

	if rank1 != rank2
        	return rank1 > rank2
    	end

	# Tie-breaking logic for equal strength hands

	# Case 1: Flush
	if rank1 in [FLUSH, STRAIGHT_FLUSH, ROYAL_FLUSH]
        	highestCard1 = get_highest_card(hand1, false)
        	highestCard2 = get_highest_card(hand2, false)

        	if highestCard1.rank != highestCard2.rank
            		return highestCard1.rank > highestCard2.rank
        	end
		return suit_rank(highestCard1.suit) > suit_rank(highestCard2.suit)
    	end

	# Case 2: Straights
	if rank1 == STRAIGHT
        	ace_low1 = is_ace_low_straight(hand1)
        	ace_low2 = is_ace_low_straight(hand2)

        	highestCard1 = get_highest_card(hand1, ace_low1)
        	highestCard2 = get_highest_card(hand2, ace_low2)

		return suit_rank(highestCard1.suit) > suit_rank(highestCard2.suit)
    	end

	# Case 3: Two Pair 
	if rank1 == TWO_PAIR
        	highPair1 = get_highest_pair_rank(hand1)
        	highPair2 = get_highest_pair_rank(hand2)

        	if highPair1 != highPair2
            		return highPair1 > highPair2
        	end

        	lowPair1 = get_lowest_pair_rank(hand1)
        	lowPair2 = get_lowest_pair_rank(hand2)

        	if lowPair1 != lowPair2
            		return lowPair1 > lowPair2
        	end

        	kicker1 = get_kicker(hand1)
        	kicker2 = get_kicker(hand2)

		return suit_rank(hand1[kicker1].suit) > suit_rank(hand2[kicker2].suit)
    	end

	# Case 4: Pair
	if rank1 == PAIR
        	pairRank1 = get_pair_rank(hand1)
        	pairRank2 = get_pair_rank(hand2)

        	if pairRank1 != pairRank2
            		return pairRank1 > pairRank2
        	end

        	highVal1 = get_highest_non_pair_card(hand1)
        	highVal2 = get_highest_non_pair_card(hand2)

        	return suit_rank(highVal1.suit) > suit_rank(highVal2.suit)
    	end

	# Case 5: High Card
	if rank1 == HIGH_CARD
        	highestCard1 = get_highest_card(hand1, false)
        	highestCard2 = get_highest_card(hand2, false)

        	if highestCard1.rank != highestCard2.rank
            		return highestCard1.rank > highestCard2.rank
        	end

        	return suit_rank(highestCard1.suit) > suit_rank(highestCard2.suit)
    	end

	if rank1 == FOUR_OF_A_KIND
        	quadRank1 = get_quad_rank(hand1)
        	quadRank2 = get_quad_rank(hand2)

        	return quadRank1 > quadRank2
    	end

	if rank1 == THREE_OF_A_KIND
        	tripletRank1 = get_triplet_rank(hand1)
        	tripletRank2 = get_triplet_rank(hand2)

        	return tripletRank1 > tripletRank2
    	end

	if rank1 == FULL_HOUSE
        	tripletRank1 = get_triplet_rank(hand1)
        	tripletRank2 = get_triplet_rank(hand2)

        	if tripletRank1 != tripletRank2
            		return tripletRank1 > tripletRank2
        	end

        	pairRank1 = get_pair_rank(hand1)
        	pairRank2 = get_pair_rank(hand2)

        	return pairRank1 > pairRank2
    	end

    	return false
end

# Function to read a deck from a file and populate hands
function read_deck_from_file(filename::String, hands::Vector{Vector{Card}})
	success = true  # Assume success unless a duplicate is found
	open(filename, "r") do inputFile
        	seenCards = Set{String}()
        	lines = readlines(inputFile)

		for line in lines
			println(line)  # Display each line to user before processing
		end

        	resize!(hands, 6)

		# Initialize each hand as an empty Vector{Card}
        	for i in 1:6
            		hands[i] = Vector{Card}()
        	end

        	for i in 1:6
        		line = lines[i]
            		cardStrs = split(line, ',')

            		for j in 1:5
                		cardStr = cardStrs[j]
                		cardStr = replace(cardStr, " " => "")  # Remove spaces inline

                		# Duplicate card check
                		if cardStr in seenCards
                    			println("\n*** ERROR - DUPLICATED CARD FOUND IN DECK ***\n\n")
                    			println("*** DUPLICATE: $cardStr ***")
                    			success = false
					break
                		end
				if !success
					break
				end

                		push!(seenCards, cardStr)

                		# Extract suit and rank
                		suit = cardStr[end]
                		rank = if cardStr[1] in '2':'9'
                    			parse(Int, cardStr[1])
                		elseif cardStr[1:2] == "10"
                    			10
                		else
                    			match = Dict('J' => 11, 'Q' => 12, 'K' => 13, 'A' => 14)
                    			match[cardStr[1]]
        	        	end

                		push!(hands[i], Card(rank, suit))
            		end
		end
	end
	if success
    		println()
    		print_six_hands(hands)
	end

	return success
end

# main method to handle overall program flow 
function main(args::Vector{String})
    	println("*** P O K E R    H A N D    A N A L Y Z E R ***\n\n")

    	deck = Deck()
	hands = Vector{Card}[] 
	resize!(hands, 6) 
	handRanks = HandRank[]

	if length(args) > 0  # Part 2: With command-line arguments
		println("*** USING TEST DECK ***")
		println("\n*** File: ", args[1])
		if !read_deck_from_file(args[1], hands)
			return
		end
	else  # Part 1: Without command-line arguments
		println("*** USING RANDOMIZED DECK OF CARDS ***")
    		shuffle_deck!(deck)
    		println("\n*** Shuffled Deck of Cards ***")
    		print_deck(deck, false) 
		hands = deal_hands(deck, 6, 5)
		print_six_hands(hands)
		println("*** Here is what remains in the deck...")
		print_deck(deck, true)
		println()
	end

	sorted_hands = sort_hands(hands, handRanks)
    	print_winning_hand(sorted_hands, handRanks)
end

main(ARGS)
